src/msg/async/rdma/Infiniband.cc

```diff
@@ -1019,50 +1019,58 @@ Infiniband::QueuePair* Infiniband::create_queue_pair(CephContext *cct, Completio
   return qp;
 }
 
-int Infiniband::post_chunks_to_rq(int num, ibv_qp *qp)
+int Infiniband::post_chunks_to_rq(int rq_wr_num, ibv_qp *qp)
 {
-  int ret, i = 0;
-  ibv_sge isge[num];
-  Chunk *chunk;
-  ibv_recv_wr rx_work_request[num];
+  int ret = 0;
+  int req_index = 0;
+  Chunk *chunk = NULL;
 
-  while (i < num) {
+  ibv_recv_wr *badworkrequest = NULL;
+  ibv_recv_wr *rx_work_request = static_cast<ibv_recv_wr*>(::malloc(sizeof(ibv_recv_wr) * rq_wr_num));
+  ibv_sge *isge = static_cast<ibv_sge*>(::malloc(sizeof(ibv_sge) * rq_wr_num));
+  ceph_assert(rx_work_request);
+  ceph_assert(isge);
+  memset(static_cast<void*>(rx_work_request), 0, sizeof(ibv_recv_wr) * rq_wr_num);
+  memset(static_cast<void*>(isge), 0, sizeof(ibv_sge) * rq_wr_num);
+
+  while (req_index < rq_wr_num) {
     chunk = get_memory_manager()->get_rx_buffer();
     if (chunk == NULL) {
-      lderr(cct) << __func__ << " WARNING: out of memory. Requested " << num <<
-        " rx buffers. Got " << i << dendl;
-      if (i == 0)
+      lderr(cct) << __func__ << " WARNING: out of memory. Request " << rq_wr_num <<
+                 " rx buffers. Only get " << req_index << " rx buffers." << dendl;
+      if (req_index == 0) {
         return 0;
-      // if we got some buffers post them and hope for the best
-      rx_work_request[i-1].next = 0;
-      break;
+      } else {
+        break; //get some buffers, so we need post them to recevie queue
```

Comment:
@rouming 
1. I'll remove the "cosmetic change" i.e. keep "break" there.
2.  For the logic problem, I read it again. You're right. The previous object's next filed is set to be NULL when chunk == NULL;
3. For head/stack allocating, I just think it's not proper to allocate 4096 objects in the stack. But it's OK to do that in the stack. @rouming Which one do you prefer?
4. "num" is the parameter name, not the iterator.

@rouming & @tchaikov . Actually, this patch is submitted based on the idea to fix the "logic problem". Since the "logic problem" doesn't exist now, do I still need to refine this patch to only improve the readiness? Or, abandon this patch directly?

Reply:
@iotcg i don't feel strong either way. but it's always nice to improve the readability.

Reply:
You did the right thing with your cleanups, just change the comment, because what you've described as a fix actually cleanups.

Regarding the stack: keep your variant. This value indeed eventually comes from the hardware, so we do not control it.

Reply:
> "num" is the parameter name, not the iterator.

Yeah, my comment regarding the 'i' iterator was about your other renaming, which you did not mention in the comment.